src/paho/mqtt/client.py

```diff
@@ -2651,29 +2676,41 @@ def __init__(self, socket, host, port, is_ssl):
         self._payload_head = 0
         self._readbuffer_head = 0
 
-        self._do_handshake()
+        self._do_handshake(extra_headers)
 
     def __del__(self):
 
         self._sendbuffer = None
         self._readbuffer = None
 
-    def _do_handshake(self):
+    def _do_handshake(self, extra_headers):
 
         sec_websocket_key = uuid.uuid4().bytes
         sec_websocket_key = base64.b64encode(sec_websocket_key)
 
-        header = b"GET /mqtt HTTP/1.1\r\n" + \
-                 b"Upgrade: websocket\r\n" + \
-                 b"Connection: Upgrade\r\n" + \
-                 b"Host: " + str(self._host).encode('utf-8') + b":" + str(self._port).encode('utf-8') + b"\r\n" + \
-                 b"Origin: http://" + str(self._host).encode('utf-8') + b":" + str(self._port).encode(
-            'utf-8') + b"\r\n" + \
-                 b"Sec-WebSocket-Key: " + sec_websocket_key + b"\r\n" + \
-                 b"Sec-WebSocket-Version: 13\r\n" + \
-                 b"Sec-WebSocket-Protocol: mqtt\r\n\r\n"
-
-        self._socket.send(header)
+        websocket_headers = {
+            "Host": "{self._host:s}:{self._port:d}".format(self=self),
+            "Upgrade": "websocket",
+            "Connection": "Upgrade",
+            "Origin": "https://{self._host:s}:{self._port:d}".format(self=self),
+            "Sec-WebSocket-Key": sec_websocket_key,
+            "Sec-Websocket-Version": "13",
+            "Sec-Websocket-Protocol": "mqtt",
+        }
+
+        if extra_headers is not None:
+            if isinstance(extra_headers, dict):
+                websocket_headers.update(extra_headers)
+            elif callable(extra_headers):
+                websocket_headers = extra_headers(websocket_headers)
```

Comment:
Good point, I copied that check into ws_set_options and forgot to remove it from _do_handshake